src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,768 @@
+
+* **Preserve**:
+  * **On Serialize**: When writing complex types, the serializer also writes them metadata ($id, $values and $ref) properties in order re-use them by writing a reference to the object or array.
+  * **On Deserialize**: While the other options show no effect on Deserialization, `Preserve` does affect its behavior with the following: Metadata will be expected (although is not mandatory) and the deserializer will try to understand it. 
+
+For System.Text.Json, the goal is to stick to the same *metadata* semantics for preserve from Newtonsoft.Json and provide a similar usage in `JsonSerializerOptions` that encompasses the needed options (i.e. provide reference preservation).
+
+This API is exposing the `ReferenceHandling` property as a class, to be extensible in the future; and provide built-in static instances of `Default` and `Preserve` that are useful to enable the most common behaviors by just setting those in `JsonSerializerOptions.ReferenceHandling`.
+
+With `ReferenceHandling` being a class, we can exclude things that, as of now, we are not sure are required and add them later based on customer feedback. For example, the `Object` and `Array` granularity of `Newtonsoft.Json's` `PreserveReferencesHandling` or the `ReferenceLoopHandling.Ignore` option.
+
+## Future
+
+Things that may build on top based on customer feedback:
+
+* (De)Serialize can define its own Preserve References Handling behavior each (i.e: you could opt-out form Preserve Reference on serialization but opt-in for read them on deserialization).
+
+* Expose a `ReferenceResolver` to override the logic that preserves references (Create your own implementation of a reference resolver).
+
+* Expose the `ReferenceResolver` in Converters to have access to the map of references.
+
+* Create `JsonReferenceHandlingAttribute` to enable to annotate properties and classes with their own isolated ReferenceHandling behavior (I cut-off the support for this due the constant checking for attributes was causing too much perf overhead on the main path, maybe we can try moving the attribute check to the warm-up method to reduce the runtime increase).
+```cs
+// Example of a class annotated with JsonReferenceHandling attributes.
+[JsonReferenceHandling(ReferenceHandling.Preserve)]
+public class Employee { 
+    string Name { get; set; }
+
+    [JsonReferenceHandling(ReferenceHandling.Ignore)]
+    Employee Manager { get; set; }
+    
+    List<Employee> Subordinates { get; set; }
+}
+```
+
+## Compatibility
+
+The next table show the combination of Newtonsoft's **ReferenceLoopHandling** and **PreserveReferencesHandling** and how to get its equivalent on System.Text.Json's *ReferenceHandling*:
+
+|       RLH\PRH |      None |              All |          Objects |           Arrays |
+|--------------:|----------:|-----------------:|-----------------:|-----------------:|
+|     **Error** | *Default* | future (overlap) | future (overlap) | future (overlap) |
```

Comment:
Is overlap the right term here? What is the Newtonsoft.Json behavior for all these permutation? They aren't really undefined. Maybe say: low usage/awkward semantics?